File: src/libs/rust/m3/src/tiles/tile.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use core::fmt;

use crate::cap::{CapFlags, Capability, Selector};
use crate::errors::{Code, Error};
use crate::kif::TileDesc;
use crate::quota::Quota;
use crate::rc::Rc;
use crate::syscalls;
use crate::tcu::TileId;
use crate::tiles::Activity;

/// Represents a tile in the tiled architecture.
pub struct Tile {
    cap: Capability,
    id: TileId,
    desc: TileDesc,
    free: bool,
}

/// Contains the different quotas for a tile
#[derive(Default)]
pub struct TileQuota {
    eps: Quota<u32>,
    time: Quota<u64>,
    pts: Quota<usize>,
}

impl TileQuota {
    /// Creates a new `TileQuota` object from given quotas.
    pub fn new(eps: Quota<u32>, time: Quota<u64>, pts: Quota<usize>) -> Self {
        Self { eps, time, pts }
    }

    /// Returns the endpoint quota
    pub fn endpoints(&self) -> &Quota<u32> {
        &self.eps
    }

    /// Returns the time quota
    pub fn time(&self) -> &Quota<u64> {
        &self.time
    }

    /// Returns the page-table quota
    pub fn page_tables(&self) -> &Quota<usize> {
        &self.pts
    }
}

impl fmt::Debug for TileQuota {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        write!(
            f,
            "TileQuota[eps={}, time={}, pts={}]",
            self.endpoints(),
            self.time(),
            self.page_tables()
        )
    }
}

impl Tile {
    /// Allocates a new tile from the resource manager with given description
    pub fn new(desc: TileDesc) -> Result<Rc<Self>, Error> {
        let sel = Activity::own().alloc_sel();
        let (id, ndesc) = Activity::own().resmng().unwrap().alloc_tile(sel, desc)?;
        Ok(Rc::new(Tile {
            cap: Capability::new(sel, CapFlags::KEEP_CAP),
            id,
            desc: ndesc,
            free: true,
        }))
    }

    /// Binds a new tile object to given selector
    pub fn new_bind(id: TileId, desc: TileDesc, sel: Selector) -> Self {
        Tile {
            cap: Capability::new(sel, CapFlags::KEEP_CAP),
            id,
            desc,
            free: false,
        }
    }

    /// Gets a tile with given description.
    ///
    /// The description is an '|' separated list of properties that will be tried in order. Two
    /// special properties are supported:
    /// - "own" to denote the own tile (provided that it has support for multiple activities)
    /// - "clone" to denote a separate tile that is identical to the own tile
    ///
    /// For other properties, see [`TileDesc::with_properties`].
    ///
    /// Examples:
    /// - tile with an arbitrary ISA, but preferred the own: "own|core"
    /// - Identical tile, but preferred a separate one: "clone|own"
    /// - BOOM core if available, otherwise any core: "boom|core"
    /// - BOOM with NIC if available, otherwise a Rocket: "boom+nic|rocket"
    pub fn get(desc: &str) -> Result<Rc<Self>, Error> {
        let own = Activity::own().tile();
        for props in desc.split('|') {
            match props {
                "own" => {
                    if own.desc().supports_tilemux() && own.desc().has_virtmem() {
                        return Ok(own.clone());
                    }
                },
                "clone" => {
                    if let Ok(tile) = Self::new(own.desc()) {
                        return Ok(tile);
                    }
                },
                p => {
                    let base = TileDesc::new(own.desc().tile_type(), own.desc().isa(), 0);
                    if let Ok(tile) = Self::new(base.with_properties(p)) {
                        return Ok(tile);
                    }
                },
            }
        }
        Err(Error::new(Code::NotFound))
    }

    /// Derives a new tile object from `self` with a subset of the resources, removing them from
    /// `self`
    ///
    /// The three resources are the number of EPs, the time slice length in nanoseconds, and the
    /// number of page tables.
    pub fn derive(
        &self,
        eps: Option<u32>,
        time: Option<u64>,
        pts: Option<u64>,
    ) -> Result<Rc<Self>, Error> {
        let sel = Activity::own().alloc_sel();
        syscalls::derive_tile(self.sel(), sel, eps, time, pts)?;
        Ok(Rc::new(Tile {
            cap: Capability::new(sel, CapFlags::empty()),
            desc: self.desc(),
            id: self.id(),
            free: false,
        }))
    }

    /// Returns the selector
    pub fn sel(&self) -> Selector {
        self.cap.sel()
    }

    /// Returns the tile id
    pub fn id(&self) -> TileId {
        self.id
    }

    /// Returns the tile description
    pub fn desc(&self) -> TileDesc {
        self.desc
    }

    /// Returns the EP, time, and page table quota
    pub fn quota(&self) -> Result<TileQuota, Error> {
        syscalls::tile_quota(self.sel())
    }

    /// Sets the quota of the tile with given selector to specified initial values (given time slice
    /// length and number of page tables).
    ///
    /// This call requires a root tile capability.
    pub fn set_quota(&self, time: u64, pts: u64) -> Result<(), Error> {
        syscalls::tile_set_quota(self.sel(), time, pts)
    }
}

impl Drop for Tile {
    fn drop(&mut self) {
        if self.free {
            Activity::own().resmng().unwrap().free_tile(self.sel()).ok();
        }
    }
}

impl fmt::Debug for Tile {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        write!(
            f,
            "Tile{}[sel: {}, desc: {:?}]",
            self.id(),
            self.sel(),
            self.desc()
        )
    }
}

File: src/libs/m3/tiles/Tile.cc
/*
 * Copyright (C) 2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/Exception.h>
#include <m3/Syscalls.h>
#include <m3/session/ResMng.h>
#include <m3/tiles/Activity.h>
#include <m3/tiles/Tile.h>

namespace m3 {

constexpr size_t MAX_DESC_LEN = 256;

static TileDesc desc_with_properties(TileDesc desc, const char *props) {
    char props_cpy[MAX_DESC_LEN];
    if(strlen(props) >= MAX_DESC_LEN)
        vthrow(Errors::NO_SPACE, "Tile description too long"_cf);
    strcpy(props_cpy, props);

    auto res = desc;
    char *prop = strtok(props_cpy, "+");
    while(prop != nullptr) {
        if(strcmp(prop, "arm") == 0)
            res = TileDesc(res.type(), TileISA::ARM, 0);
        else if(strcmp(prop, "x86") == 0)
            res = TileDesc(res.type(), TileISA::X86, 0);
        else if(strcmp(prop, "riscv") == 0)
            res = TileDesc(res.type(), TileISA::RISCV, 0);
        else if(strcmp(prop, "rocket") == 0)
            res = TileDesc(res.type(), res.isa(), 0, res.attr() | TileAttr::ROCKET);
        else if(strcmp(prop, "boom") == 0)
            res = TileDesc(res.type(), res.isa(), 0, res.attr() | TileAttr::BOOM);
        else if(strcmp(prop, "nic") == 0)
            res = TileDesc(res.type(), res.isa(), 0, res.attr() | TileAttr::NIC);
        else if(strcmp(prop, "serial") == 0)
            res = TileDesc(res.type(), res.isa(), 0, res.attr() | TileAttr::SERIAL);
        else if(strcmp(prop, "kecacc") == 0)
            res = TileDesc(res.type(), res.isa(), 0, res.attr() | TileAttr::KECACC);
        else if(strcmp(prop, "indir") == 0)
            res = TileDesc(TileType::COMP, TileISA::ACCEL_INDIR, 0, TileAttr::IMEM);
        else if(strcmp(prop, "copy") == 0)
            res = TileDesc(TileType::COMP, TileISA::ACCEL_COPY, 0, TileAttr::IMEM);
        else if(strcmp(prop, "rot13") == 0)
            res = TileDesc(TileType::COMP, TileISA::ACCEL_ROT13, 0, TileAttr::IMEM);
        else if(strcmp(prop, "idedev") == 0)
            res = TileDesc(TileType::COMP, TileISA::IDE_DEV, 0, TileAttr::IMEM);
        else if(strcmp(prop, "nicdev") == 0)
            res = TileDesc(TileType::COMP, TileISA::NIC_DEV, 0, TileAttr::IMEM);
        else if(strcmp(prop, "serdev") == 0)
            res = TileDesc(TileType::COMP, TileISA::SERIAL_DEV, 0, TileAttr::IMEM);
        prop = strtok(NULL, "+");
    }
    return res;
}

Tile::~Tile() {
    if(_free) {
        try {
            Activity::own().resmng()->free_tile(sel());
        }
        catch(...) {
            // ignore
        }
    }
}

Reference<Tile> Tile::alloc(const TileDesc &desc, bool init) {
    capsel_t sel = SelSpace::get().alloc_sel();
    TileDesc res = Activity::own().resmng()->alloc_tile(sel, desc, init);
    return Reference<Tile>(new Tile(sel, res, KEEP_CAP, true));
}

Reference<Tile> Tile::get(const char *desc, bool init) {
    char desc_cpy[MAX_DESC_LEN];
    if(strlen(desc) >= MAX_DESC_LEN)
        vthrow(Errors::NO_SPACE, "Properties description too long"_cf);
    strcpy(desc_cpy, desc);

    auto own = Activity::own().tile();
    char *props = strtok(desc_cpy, "|");
    while(props != nullptr) {
        if(strcmp(props, "own") == 0) {
            if(own->desc().supports_tilemux() && own->desc().has_virtmem())
                return own;
        }
        else if(strcmp(props, "clone") == 0) {
            // on m3lx, we don't support "clone", because the required semantics are difficult to
            // support. At first, being a clone requires to have the same multiplexer type, i.e.,
            // Linux again. And the semantics of Tile::get("clone") are that we get a new tile for
            // ourself, which would require us to boot up a new Linux instance. This takes simply
            // too long to do that dynamically, IMO. Therefore, the most sensible way to handle
            // "clone" on m3lx is to let it always fail. Meaning, applications should provide "own"
            // as a fallback.
#if !defined(__m3lx__)
            try {
                return Tile::alloc(own->desc(), init);
            }
            catch(...) {
            }
#endif
        }
        else if(strcmp(props, "compat") == 0) {
            // same as for "clone"
#if !defined(__m3lx__)
            try {
                auto type_isa = TileDesc(own->desc().type(), own->desc().isa(), 0);
                return Tile::alloc(type_isa, init);
            }
            catch(...) {
            }
#endif
        }
        else {
            auto base = TileDesc(own->desc().type(), own->desc().isa(), 0);
            try {
                return Tile::alloc(desc_with_properties(base, props), init);
            }
            catch(...) {
            }
        }
        props = strtok(NULL, "|");
    }
    vthrow(Errors::NOT_FOUND, "Unable to find tile with {}"_cf, desc);
}

Reference<Tile> Tile::derive(Option<uint> eps, Option<TimeDuration> time, Option<size_t> pts) {
    capsel_t sel = SelSpace::get().alloc_sel();
    Syscalls::derive_tile(this->sel(), sel, eps, time, pts);
    return Reference<Tile>(new Tile(sel, desc(), 0, false));
}

KIF::Syscall::MuxType Tile::mux_type() const {
    return Syscalls::tile_mux_info(sel());
}

std::tuple<Quota<uint>, Quota<TimeDuration>, Quota<size_t>> Tile::quota() const {
    return Syscalls::tile_quota(sel());
}

void Tile::set_quota(TimeDuration time, size_t pts) {
    Syscalls::tile_set_quota(sel(), time, pts);
}

}
